goto_checkt: keep same-assertion for different expressions

tautschnig opened the pull request:
~[Only the second commit is relevant, the first one is #5844 and only required to have the necessary test in place.]~

Purging redundant assertions results in surprising user output as it would seem that certain properties aren't being checked.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
goto_checkt: keep same-assertion for different expressions

Purging redundant assertions results in surprising user output as it
would seem that certain properties aren't being checked.

martin-cs reviewed (approved):
I'm going with a tentative 'yes'.  Assorted thoughts:

- When doing work on how to QA / evaluate / certify program analysis tools, I came to the conclusion that omitting VCs was actually quite a bit worse than simply getting them wrong.  If you have a VC and a status then you have a reasonable chance of being able to check that it is correct; there are a bunch of techniques.  If you don't have the VC there are many fewer ways of finding that it is missing.  As such I am suspicious of anything that removes VCs, even the ones that are "obviously" safe or redundant.
- I am working with a colleague on tracking down missing VCs.  This is a much more tedious and harder process than it should be.  Let's not have concealed cracks where VCs can just vanish.
- If a VC is trivial then we should still report it, we just need to make sure that it is minimal overhead.

tautschnig commented:
@martin-cs Thank you very much for your elaboration, I fully concur. It seems a further fix is in order: d974aa38 introduced a "retain-trivial" option, but this option was never made available in any command-line parser...

tautschnig committed:
Permit keeping trivially-true assertions

In some settings it is desirable to obtain a report of all properties
being checked, including those that are trivially true. There used to be
an option "all-claims", then renamed to "retain-trivial", but the latter
was never made available in command-line front-ends.

The option is now renamed to retain-trivial-checks, and made available
in all front-ends.

tautschnig commented:
> @martin-cs Thank you very much for your elaboration, I fully concur. It seems a further fix is in order: [d974aa3](https://github.com/diffblue/cbmc/commit/d974aa3897271d8226f1cd82bb3d279ac665f4a8) introduced a "retain-trivial" option, but this option was never made available in any command-line parser...

Such a fix is now included as well.
